io.openems.edge.evcs.goe.chargerhome/src/io/openems/edge/evcs/goe/chargerhome/GoeAPI.java

```diff
@@ -0,0 +1,259 @@
+
+import java.io.BufferedReader;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.net.HttpURLConnection;
+import java.net.URL;
+
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+
+import io.openems.common.exceptions.OpenemsError.OpenemsNamedException;
+import io.openems.common.exceptions.OpenemsException;
+import io.openems.common.utils.JsonUtils;
+
+public class GoeAPI {
+	
+	private final String IP;
+	private boolean debugMode;
+	private String debugFile;
+	private int executeEveryCycle;
+	private int cycle;
+	private JsonObject jsonStatus;
+	private GoeChargerHomeImpl parent;
+
+	public GoeAPI(String ip, boolean debugMode, String debugFile, int StatusAfterCycles, GoeChargerHomeImpl p) {
+		this.IP = ip;
+		this.debugMode = debugMode;
+		this.debugFile = debugFile;
+		this.cycle = 0;
+		this.executeEveryCycle = StatusAfterCycles;
+		this.jsonStatus = null;
+		this.parent = p;
+	}
+
+	/**
+	 * Gets the status from go-e
+	 * 
+	 * See https://github.com/goecharger
+	 * 
+	 * @return the boolean value
+	 * @throws OpenemsNamedException on error
+	 */
+	public JsonObject getStatus() {
+			
+		try {
+			// Execute every x-Cycle
+			if (this.cycle == 0 || this.cycle % this.executeEveryCycle == 0) {
+				JsonObject json = new JsonObject();
+				if (!debugMode) {
+					String URL = "http://" + this.IP + "/status";
+					json = this.sendRequest(URL, "GET");
+				}						
+			
+				if (debugMode) {
+					JsonParser parser = new JsonParser();
+					Object obj = parser.parse(new FileReader(debugFile));
+					json = (JsonObject) obj;
+				}										
+				
+				this.cycle = 1;
+				this.jsonStatus = json;
+				return json;		
+			}
+			else {
+				this.cycle++;
+				return this.jsonStatus;
+			}
+
+		} catch (Exception e) {
+			e.printStackTrace();
+			return null;
+		}
+	}
+	
+	/** Sets the activation status for go-e
+	 * 
+	 * See https://github.com/goecharger
+	 * 
+	 * @return JsonObject with new settings
+	 * @throws OpenemsNamedException on error
+	 */
+	public JsonObject setActive(boolean active) {
+			
+		try {
+			if (active != this.parent.Active) {
+				JsonObject json = new JsonObject();
+				if (!debugMode) {
+					Integer status = 0;
+					if (active) {
+						status = 1;
+					}
+					String URL = "http://" + this.IP + "/mqtt?payload=alw=" + Integer.toString(status);
+					json = this.sendRequest(URL, "PUT");
+					this.parent.Active = active;					
+					return json;
+				}		
+				else {
+					this.parent.Active = active;
+					return this.getStatus();
+				}
+			}
+			else {
+				return this.jsonStatus;
+			}			
+
+		} catch (Exception e) {
+			e.printStackTrace();
+			return null;
+		}
+	}
+	
+	/** Sets the Current in Ampere for go-e
+	 * 
+	 * See https://github.com/goecharger
+	 * 
+	 * @return JsonObject with new settings
+	 * @throws OpenemsNamedException on error
+	 */
+	public JsonObject setCurrent(int current) {
```

Comment:
Maybe you could add a check if the entered current is greater than or equal to the minimum current and less than or equal to the maximum current. (I don't know but maybe you have built in that the setCurrent function can only be called with values between min and max. Then the recheck would be again pointless)

Code in Java for smth like that

```
public static int constrain_value(int value, int min, int max) { 
   if(min > max){
      //Switch Min and Max if Min is bigger than max
      return constrain_value(value, max, min);
   }else{
      //Normal Return Path
      if(min == max){
         //If Min and Max are the Same the Value must also be the Same so return min or max 
         return min;
      }else{
         if(value >= min){
            if(value <= max){
               return value;
            }else{
               return max;
            }  
         }else{
            return min;
         }
      }
   }   
}
```

Reply:
This is already done in when the method is called - could be done in here as well.